frame/data-feed/src/lib.rs

```diff
@@ -0,0 +1,704 @@
+//!     fn get() -> u32 {
+//!         // it's a way to convert DataType to expected type
+//!         match StorageArgument::get() {
+//!             pallet_data_feed::DataType::U128(d) => d.saturated_into::<u32>(),
+//!             _ => 0, // return a default value
+//!         }
+//!     }
+//! }
+//! impl template::Trait for Runtime {
+//!     type Event = Event;
+//!     type Data = SimpleDataFeedGet;
+//! }
+//! // in data-feed module
+//! // 1. first need to register DataProvider.
+//! // 2. then need to register StorageKey:
+//! //      1. use extrinsic to call `register_storage_key`
+//! //      2. the param: `key: StorageKey` is a bytes from template module using `:StorageArgument:`
+//! //      (which is generated from macro `parameter_types`, if developer use other method for storage, use that storage key)
+//! //      3. the param: `info: Info<T::BlockNumber>` would contain operation type, schedule block interval and so no.
+//! // 3. use offchain worker or a program beside the chain to submit `feed_data` extrinsic to feed data.
+//! // 4. template module could use T::Data::get() to use the feeded data.
+//! ```
+//!
+//! ## Genesis config
+//!
+//! ## Assumptions
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod tests;
+
+use codec::{Decode, Encode};
+use lite_json::json::{JsonValue, NumberValue};
+
+use sp_core::crypto::KeyTypeId;
+use sp_runtime::{
+	offchain::{http, Duration},
+	traits::{CheckedDiv, Saturating, StaticLookup, Zero},
+	DispatchError, DispatchResult, FixedPointNumber, FixedU128, RuntimeDebug,
+};
+use sp_std::{prelude::*, str::Chars};
+
+use frame_support::{
+	debug, decl_error, decl_event, decl_module, decl_storage,
+	traits::{Contains, EnsureOrigin},
+	IterableStorageDoubleMap, IterableStorageMap,
+};
+use frame_system::{
+	ensure_root, ensure_signed,
+	offchain::{AppCrypto, CreateSignedTransaction, SendSignedTransaction, Signer},
+};
+
+/// The identifier for data-feed-specific crypto keys
+pub const KEY_TYPE: KeyTypeId = KeyTypeId(*b"dafe");
+pub const RING_BUF_LEN: usize = 8;
+pub const MAX_KEY_LEN: usize = 8;
+pub const MAX_PROVIDER_LEN: usize = 8;
+pub type StorageKey = Vec<u8>;
+
+pub mod crypto {
+	use super::KEY_TYPE;
+	use sp_core::sr25519::Signature as Sr25519Signature;
+	use sp_runtime::{
+		app_crypto::{app_crypto, sr25519},
+		traits::Verify,
+	};
+	app_crypto!(sr25519, KEY_TYPE);
+
+	pub type AuthoritySignature = Signature;
+	pub type AuthorityId = Public;
+
+	pub struct DataFeedAuthId;
+	impl frame_system::offchain::AppCrypto<<Sr25519Signature as Verify>::Signer, Sr25519Signature>
+		for DataFeedAuthId
+	{
+		type RuntimeAppPublic = AuthorityId;
+		type GenericSignature = sp_core::sr25519::Signature;
+		type GenericPublic = sp_core::sr25519::Public;
+	}
+}
+
+/// The operation type that specifies how we calculate the data
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Encode, Decode, RuntimeDebug)]
+pub enum Operations {
+	Average,
```

Comment:
I havent read more, but median value is quite nice to have because it is resistant to outliers/extreme values.